iceoryx_hoofs/test/moduletests/test_cxx_functional_interface_and_then.cpp

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+#include "test_cxx_functional_interface_types.hpp"
+
+namespace
+{
+using namespace test_cxx_functional_interface;
+using namespace ::testing;
+
+#define IOX_TEST(TestName, variationPoint)                                                                             \
+    using SutType = typename TestFixture::TestFactoryType::Type;                                                       \
+    TestName<iox::cxx::internal::HasValueMethod<SutType>::value>::template performTest<                                \
+        typename TestFixture::TestFactoryType>([](auto& sut, auto callback) { variationPoint.and_then(callback); })
+
+constexpr bool TYPE_HAS_VALUE_METHOD = true;
+constexpr bool TYPE_HAS_NO_VALUE_METHOD = false;
+
+template <bool HasValue>
+struct AndThenIsCalledCorrectlyWhenValid;
```

Comment:
Would it be possible to add a static_assert with a reasonable error massage that the type has not the appropriate methods?
I think it should be possible like here
```cpp
template <bool HasValue, typename SutType>
struct AndThenIsCalledCorrectlyWhenValid {
    static_assert(always_false_v(SutType) && "The type does not implement a `value` method!");
};
```

Reply:
I think this does not make sense or maybe I misunderstood you.
It is completely valid for a type to not implement a value method. The test if the `and_then` and `or_else` signature is correct has to be done then in the types tests for instance in the optional tests.

Reply:
I am confused?! I think this static assert would never been called since it cannot happen that a sut without value method is called in such a struct. I would use the same means in the static assert as I would use in the specialization of the struct. `HasValueMethod<SutType>::value`